### PR TITLE
Feature/planetexpress/psa 25834 improve netboot stack

### PR DESCRIPTION
**Pull request by mazlumtoprak**
This PR represents our new cachingserver architecture, that will be deployed with ansible. We've removed unused folders / scripts and are providing the docker images now within a public accessible registry.

changed services

- netboot-tftp (now inlcudes initial bootloaders)
- netboot-cleaner (migrated cleaner to golang)
- netboot-monitoring ( adjusted to new setup )
- netboot-sync (sync squashFS files from azure blobstorage)
- netboot-ipxeMenuGenerator (migrated to to golang by @ccadruvi )
- netboot-http (as is, available on public repo)

### CODE REVIEW

**Comment by mazlumtoprak** [comment]
**Note:** Before merging this PR, we need to make sure that we can replace all locations with either the central-cachingserver in azure or replace the current ones with the new ansible setup

File: netboot-services/ipxeMenuGenerator/main.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+func renderAdvancedMenu(filename string, netbootServerIP string) {
+	prodImages := getProdImages()
+	devImages := getDevImages()
+	j2, err := jinja2.NewJinja2("advancedmenu.ipxe", 1,
+		jinja2.WithGlobal("netbootServerIP", netbootServerIP),
+		jinja2.WithGlobal("prod", prodImages),
+		jinja2.WithGlobal("dev", devImages),
+	)
+	if err != nil {
+		panic(err)
```
Comment (ccadruvi) [inline netboot-services/ipxeMenuGenerator/main.go:169]: Is this service still functional, if we can't render the advanced menu? I would argue that it is somewhat functional, so we shouldn't panic here

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 [hunk 2]
```diff
@@ -34,19 +26,15 @@ item dg-thinclient-prod ${sp} DG ThinClient
 item --gap Erweitert:
 item advanced ${sp} Erweiterte Bootoptionen
 item reboot ${sp} Neustart
-item --gap Aktuell gesetzter Bootserver: ${next-server}
+item --gap Aktuell gesetzter Bootserver: {{ netbootServerIP }}
```
Comment (ccadruvi) [inline netboot-services/ipxeMenuGenerator/menu.ipxe.j2:29]: Why this change? We still set next-server

File: netboot-services/monitoring/monitor_tftp.sh [hunk 3]
```diff
@@ -9,7 +9,11 @@ function tftpConnectWithEchoInfluxOutput() {
         tftpHostName="$1"
         tftpServerTestFile="$2"
 
-        atftp -g -r "$tftpServerTestFile" -l /tmp/"$tftpServerTestFile" "$tftpHostName" >/dev/null 2>&1
+        if [[ $tftpServerTestFile == "menu.ipxe"  ]]; then
```
Comment (ccadruvi) [inline netboot-services/monitoring/monitor_tftp.sh:12]: Why this distinction?

File: netboot-services/sync/sync.env [hunk 4]
```diff
@@ -0,0 +1,5 @@
+SYNC_BLOB_URL="{{ sync_blob_url }}"
```
Comment (ccadruvi) [inline netboot-services/sync/sync.env:1]: Isn't this file distributed with the ansible config now?

File: netboot-services/sync/syncer.sh [hunk 5]
```diff
@@ -1,100 +1,34 @@
...
-      fi
-    done <<<"${cachingServers[@]}" # end caching for loop
-  done                             # end while loop
+while true
+do
+  if [ "$SYNC_DEV" ]; then
+    azcopy cp --cap-mbps "$SYNC_BANDWITDH_LIMIT_MBITS" "$SYNC_BLOB_URL/dev/*$SYNC_SAS_TOKEN" "/home/syncer/dev/" --overwrite=ifSourceNewer --recursive
+  fi
+ 
+  azcopy cp --cap-mbps "$SYNC_BANDWITDH_LIMIT_MBITS" "$SYNC_BLOB_URL/prod/*$SYNC_SAS_TOKEN" "/home/syncer/prod/" --overwrite=ifSourceNewer --recursive
```
Comment (ccadruvi) [inline netboot-services/sync/syncer.sh:26]: we should use the SYNC_PROD env var here or (preferably) remove the variable

**Review by ccadruvi — CHANGES_REQUESTED** [review]

File: docker-compose.yaml [hunk 6]
```diff
@@ -21,77 +23,42 @@ services:
     restart: unless-stopped
 
   netboot-cleaner:
-    image: anymodconrst001dg.azurecr.io/planetexpress/netboot-cleaner:latest
+    image: dgpublicimagesprod.azurecr.io/planetexpress/netboot-cleaner:latest
     container_name: netboot-cleaner
```
[the remark this answers is not in the document]
Comment (mazlumtoprak) [reply docker-compose.yaml:27]: nice catch

File: docker-compose.yaml [hunk 7]
```diff
@@ -1,16 +1,18 @@
-version: "3"
```
Comment (SwissGipfel) [inline docker-compose.yaml:1]: General question, not really part of your change (maybe).
we used to write mac-ipxe files directly on our main netboot server. And since every container on the netboot server ran as root, there was no problem in reading the files in the mounted folder realtime.

But now, we don't run every container as root anymore. Does that mean, that the described behavior will also change?

**Review by SwissGipfel — CHANGES_REQUESTED** [review]

File: netboot-services/cleaner/cleaner.env [hunk 8]
```diff
@@ -0,0 +1,4 @@
+THRESHOLD_MAX_IMAGES_COUNT_DEV=
```
[the remark this answers is not in the document]
Comment (mazlumtoprak) [reply netboot-services/cleaner/cleaner.env:1]: Let's go for the default values, where possible

File: netboot-services/cleaner/main.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+	"strings"
+	"syscall"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type image struct {
+	ImageName         string `json:"imageName"`
+	KernelVersionFile string `json:"KernelVersionFile"`
```
[the remark this answers is not in the document]
Comment (mazlumtoprak) [reply netboot-services/cleaner/main.go:19]: Removed json as a whole

File: netboot-services/cleaner/Dockerfile [hunk 10]
```diff
@@ -1,4 +1,10 @@
-FROM alpine:3.14.0
-RUN apk add --no-cache bash
-COPY cleaner.sh /cleaner.sh
-CMD ["/cleaner.sh"]
+FROM golang:1.20.5 AS build
```
[the remark this answers is not in the document]
Comment (mazlumtoprak) [reply netboot-services/cleaner/Dockerfile:1]: Adjusted

**Comment by mazlumtoprak** [comment]
Thanks @SwissGipfel & @ccadruvi for the review. I've applied a lot of changes which we mostly discussed bilateral. In my view, we should be good to merge. Only waiting for https://github.com/DigitecGalaxus/netboot/pull/10#issuecomment-1621688365 .

**Review by SwissGipfel — DISMISSED** [review]

**Review by SwissGipfel — APPROVED** [review]
